### ISSUE TITLE
Plotting a Map on an Axes with a different projection is (silently) wrong

### ISSUE THREAD

**Issue by dstansby** [sunpy/sunpy#5163]
Below is an example that plots two maps. In the first case, everything is normal (`projection=map_aia`, plotting `map_aia`), but in the second case I use `projection=map_aia`, and plot `map_hmi`. Because HMI has ~180deg rotation with respect to AIA, this silently leads to an incorrect plot. Notice how the axes are the same, but the strong magnetic fields in HMI do not line up with the ARs in AIA:
![hmi_aia_plot](https://user-images.githubusercontent.com/6197628/113137867-ff021c00-921c-11eb-931a-d4119f74dae6.png)

We should raise a `NotImplementedError` if a user tries to do this, and eventually support doing this via. reprojection. This might really be an upstream astropy issue.

```python
import matplotlib.pyplot as plt
import sunpy.data.sample
import sunpy.map


map_aia = sunpy.map.Map(sunpy.data.sample.AIA_171_IMAGE)
map_hmi = sunpy.map.Map(sunpy.data.sample.HMI_LOS_IMAGE)
map_hmi.plot_settings['cmap'] = "hmimag"
map_hmi.plot_settings['norm'] = plt.Normalize(-1500, 1500)


fig = plt.figure(figsize=(12, 5))
ax1 = fig.add_subplot(1, 2, 1, projection=map_aia)
map_aia.plot(axes=ax1)
# Note that projection is AIA here
ax2 = fig.add_subplot(1, 2, 2, projection=map_aia)
map_hmi.plot(axes=ax2)

plt.show()
```

**Comment by nabobalis** [comment]
Was this addressed by any of the PR's leading up to 3.0?

**Comment by ayshih** [comment]
Essentially, although not the specific implementation suggested by the OP.  The example code above now produces a `SunpyUserWarning`.  Modifying the last lines of the example code to be:
```python
...
map_hmi.plot(axes=ax2, autoalign=True)
ax2.axis(ax1.axis())

plt.show()
```
gives
![Figure_1](https://user-images.githubusercontent.com/991759/116884322-9c71b680-abf4-11eb-9865-6658bb17a4ad.png)
which has the proper orientation.
